### PR TITLE
Adding TF2 models to zoo

### PR DESCRIPTION
**Pull request by Fariborzzz**
## What changes are proposed in this pull request?

The following models are added:
centernet-hg104-1024-coco-tf2
centernet-resnet101-v1-fpn-512-coco-tf2
centernet-resnet50-v2-512-coco-tf2
centernet-mobilenet-v2-fpn-512-coco-tf2
efficientdet-d0-512-coco-tf2
efficientdet-d1-640-coco-tf2
efficientdet-d2-768-coco-tf2
efficientdet-d3-896-coco-tf2
efficientdet-d4-1024-coco-tf2
efficientdet-d5-1280-coco-tf2
efficientdet-d6-1280-coco-tf2
efficientdet-d7-1536-coco-tf2
ssd-mobilenet-v2-320-coco17
ssd-mobilenet-v1-fpn-640-coco17

## How is this patch tested? If it is not, please explain why.
```
import fiftyone as fo
import fiftyone.zoo as foz

dataset = foz.load_zoo_dataset("quickstart").select_fields().clone()
view = dataset.limit(10)

# Models you added should appear here
print(foz.list_zoo_models())

model = foz.load_zoo_model("<base-name>")

view.apply_model(model, label_field="predictions")

session = fo.launch_app(view)
```
### Is this a user-facing change that should be mentioned in the release notes?

-   [X] No. You can skip the rest of this section.
-   [ ] Yes. Give a description of this change to be included in the release
        notes for FiftyOne users.

(Details in 1-2 sentences. You can just refer to another PR with a description
if this PR is part of a larger change.)

### What areas of FiftyOne does this PR affect?

-   [ ] App: FiftyOne application changes
-   [ ] Build: Build and test infrastructure changes
-   [ ] Core: Core `fiftyone` Python library changes
-   [ ] Documentation: FiftyOne documentation changes
-   [ X] Other: New models are added

### CODE REVIEW

**Comment by Fariborzzz** [comment]
Hey, this is my first commit and hope followed the guidelines correctly.
Please let me know if any change is required to follow fiftyone principles.

Kind regards
Fari

**Comment by benjaminpkane** [comment]
@Fariborzzz I haven't tested the model, but your entry looks good to me. Feel free to add more :slightly_smiling_face: 

 When your ready, just update the description so it lists every new model. Thanks

**Review by brimoor — CHANGES_REQUESTED** [review]
Hey @Fariborzzz, thanks so much for taking this on!

Things are looking good to me; only comment is that you need to include the file extension (`.tar.gz` for most of these models) when setting `base_filename` in the JSON.

Here's a snippet you can use to test the models that you're adding:

```py
import os

import fiftyone as fo
import fiftyone.zoo as foz

name = "centernet-hg104-512-coco-tf2"

foz.download_zoo_model(name)

# If the model is stored in an archive, set `delete_archive=False` in the JSON
# manifest when running this for the first time so you can get the size
model_path = foz.find_zoo_model(name)
print(os.path.getsize(model_path))

model = foz.load_zoo_model(name)

dataset = foz.load_zoo_dataset("quickstart").select_fields().clone()
view = dataset.limit(10)

view.apply_model(model, label_field="predictions")

session = fo.launch_app(view)
```

The above snippet should succeed when models are configured correctly. The above code does not currently work due to the missing file extensions in `base_filename`.

**Comment by Fariborzzz** [comment]
> Hey @Fariborzzz, thanks so much for taking this on!
> 
> Things are looking good to me; only comment is that you need to include the file extension (`.tar.gz` for most of these models) when setting `base_filename` in the JSON.
> 
> Here's a snippet you can use to test the models that you're adding:
> 
> ```python
> import os
> 
> import fiftyone as fo
> import fiftyone.zoo as foz
> 
> name = "centernet-hg104-512-coco-tf2"
> 
> foz.download_zoo_model(name)
> 
> # If the model is stored in an archive, set `delete_archive=False` in the JSON
> # manifest when running this for the first time so you can get the size
> model_path = foz.find_zoo_model(name)
> print(os.path.getsize(model_path))
> 
> model = foz.load_zoo_model(name)
> 
> dataset = foz.load_zoo_dataset("quickstart").select_fields().clone()
> view = dataset.limit(10)
> 
> view.apply_model(model, label_field="predictions")
> 
> session = fo.launch_app(view)
> ```
> 
> The above snippet should succeed when models are configured correctly. The above code does not currently work due to the missing file extensions in `base_filename`.

Thank you @brimoor and @benjaminpkane. I fixed the base_filename for the mentioned models. 
Using the snippet above, I can run the models as below: 

Screenshot from 2021-12-28 12-25-16

This test is performed on tensorflow CPU.

**Review by brimoor — COMMENTED** [review]
Awesome! 💪 

Are you planning to add more models or is this it for now?

File: fiftyone/zoo/models/manifest-tf.json [hunk 1]
```diff
@@ -1333,6 +1333,576 @@
             },
             "tags": ["detection", "coco", "tf2"],
             "date_added": "2021-01-6 11:01:34"
+        },
+        {
+            "base_name": "centernet-hg104-1024-coco-tf2",
+            "base_filename": "centernet_hg104_1024x1024_coco17_tpu-32",
```
[the remark this answers is not in the document]
Comment (Fariborzzz) [reply fiftyone/zoo/models/manifest-tf.json]: My apologies, for the typo :smile: , this is fixed in the latest commit.
I'm interested to add custom models like Yolov5, Yolov4 and other detectors with specific classes if you don't mind.

[the remark this answers is not in the document]
Comment (brimoor) [reply fiftyone/zoo/models/manifest-tf.json]: Absolutely, feel free to add more! Just let us know when you're finished and we'll merge this 🚀

[the remark this answers is not in the document]
Comment (Fariborzzz) [reply fiftyone/zoo/models/manifest-tf.json]: @brimoor  I converted yolov5 models to pb format. However, im looking for an attribute for default_deployment_config_dict to resize both dimensions of images to 640 for inference. I spent a lot of time on ETA lib, but i couldn't find such a function (Just found image_max_dim). Do you have any recommendation?

**Comment by brimoor** [comment]
> @brimoor I converted yolov5 models to pb format. However, im looking for an attribute for default_deployment_config_dict to resize both dimensions of images to 640 for inference. I spent a lot of time on ETA lib, but i couldn't find such a function (Just found image_max_dim). Do you have any recommendation?

@Fariborzzz If you want to add an arbitrary TF model stored in `.pb` format to the FiftyOne Model Zoo, then you'll have to implement an appropriate [Model](https://github.com/voxel51/fiftyone/blob/abe0725ed5ff1cda14f432645547d160d4fcc666/fiftyone/core/models.py#L1531) class that knows how to load the model, feed data to it, and convert the output to the appropriate FiftyOne `Label` format.

The [TorchImageModel](https://github.com/voxel51/fiftyone/blob/abe0725ed5ff1cda14f432645547d160d4fcc666/fiftyone/utils/torch.py#L186) is an example of a pretty generic class that provides an API for loading most Torch models. It has enough (optional) config arguments to specify things like what layer to extract as output, what pre-processing to perform, and how to convert various common output formats to FiftyOne format.

We don't have a generic class like `TorchImageModel` for TF models yet, but it would definitely be very useful to have one if you're up to it! 🤗

Instead, for historical reasons, all of the TF models in the FiftyOne Model Zoo are actually loaded via the classes in the [eta.detectors](https://github.com/voxel51/eta/tree/develop/eta/detectors) module and then exposed in FiftyOne format via the [ETAModel](https://github.com/voxel51/fiftyone/blob/abe0725ed5ff1cda14f432645547d160d4fcc666/fiftyone/utils/eta.py#L87) wrapper class. For example, there's the [YOLODetector](https://github.com/voxel51/eta/blob/17fb1148e853704872ed50e0e30c7800272b8398/eta/detectors/yolo.py#L78) class that wraps the darknet YOLOv4 models.

**tl;dr** my two recommended approaches to expose your YOLOv5 model in the FiftyOne Model Zoo are:
- Take a whack at implementing a generic `fiftyone.utils.tf.TensorFlowImageModel` class that can run the model as a special case, taking inspiration from `fiftyone.utils.torch.TorchImageModel`
- Implement a more specialized `eta.detectors.yolo.YOLOv5Detector` wrapper class in ETA, taking inspiration from `eta.detectors.yolo. YOLODetector`

**Comment by Fariborzzz** [comment]
@brimoor I think this takes some time to be implemented as I cannot find class hierarchy and preprocessing modules. Feel free to merge this commit and share any other documentation which can help me with this :smile: .

Thank you :ice_cream: .

**Review by brimoor — APPROVED** [review]
LGTM. Thanks so much for the contribution! 💪
